util/memory/meminfo.go

```diff
@@ -200,3 +185,232 @@ func InstanceMemUsed() (uint64, error) {
 	serverMemUsage.set(memoryUsage, time.Now())
 	return memoryUsage, nil
 }
+
+// MemTotalCGroup returns the total amount of RAM that TIDB is limited to use in container environment.
+func MemTotalCGroup() (uint64, error) {
+	mem, t := memLimit.get()
+	if time.Since(t) < 60*time.Second {
+		return mem, nil
+	}
+	mem, err := getMemoryLimit()
+	if err != nil {
+		return mem, err
+	}
+	if mem == maximumBytes {
+		mem = hostTotalMem
+	}
+	memLimit.set(mem, time.Now())
+	return mem, nil
+}
+
+// MemUsedCGroup returns the total used amount of RAM on this system in container environment.
+func MemUsedCGroup() (uint64, error) {
+	mem, t := memUsage.get()
+	if time.Since(t) < 500*time.Millisecond {
+		return mem, nil
+	}
+	mem, err := getMemoryUsed()
+	if err != nil {
+		return mem, err
+	}
+	memUsage.set(mem, time.Now())
+	return mem, nil
+}
+
+// getMemoryLimit reads from memory subsystem and returns TIDB's memory limit
+func getMemoryLimit() (uint64, error) {
+	if isCgroupV1 {
+		limitInBytes, err := getCgroupFileValue(memCtrName, "memory.limit_in_bytes")
+		if err != nil {
+			return maximumBytes, err
+		}
+		if limitInBytes == uint64(maximumBytes) {
+			if isHierarchical {
+				return grepCgroupFile(memCtrName, "memory.stat", "hierarchical_memory_limit")
+			}
+		}
+		return limitInBytes, nil
+	}
+	// for cgroup v2, we get memory limit directly from /sys/fs/cgroup/memory.max
```

Comment:
What will happen if tidb process is attached to a child cgroup?
For example, /sys/fs/cgroup/xx/yy/zz/memory.max. We also use the value of `/sys/fs/cgroup/memory.max`?